File: 06/byungkyu/28.cpp
class Solution {
public:
    int pi[10005]={0,};
    void kmpinit(int n,string S){
        int i,j,k,a,b,c;
        a=1;b=0;
        while(a+b<n){
            if(S[a+b]==S[b]){
                pi[a+b]=b+1;
                b++;
            }
            else{
                if(b==0)a++;
                else{
                    a+=b-pi[b-1];
                    b=pi[b-1];
                }
            }
        }
    }

    int kmpfind(string L,string S,int l,int s){
        int i,j,k,a,b;
        a=0;b=0;
        while(a<=l-s){
            if(L[a+b]==S[b]&&b<s){
                b++;
                if(b>=s)return a;
            }
            else{
                if(b==0)a++;
                else{
                    a+=b-pi[b-1];
                    b=pi[b-1];
                }
            }
        }
        return -1;
    }
    int strStr(string haystack, string needle) {
        kmpinit(needle.length(),needle);
        return kmpfind(haystack,needle,haystack.length(),needle.length());
    }
};

File: 06/taehyuk/28.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        
        def getPartialMatch(N: str) -> list:
            m = len(N)
            pi = [0] * m
            j = 0 

            for i in range(1, m):
                while (j > 0 and N[i] != N[j]):
                    j = pi[j - 1]
                
                if N[i] == N[j]:
                    j += 1
                    pi[i] = j
                else:
                    pi[i] = 0
            
            return pi
        
        def searchSubstring(a: str, b: str) -> int:
            n = len(a)
            m = len(b)
            pi = getPartialMatch(b)
            j = 0

            for i in range(n):
                while (j > 0 and a[i] != b[j]):
                    j = pi[j - 1]

                if a[i] == b[j]:
                    j += 1
                
                if j == m:
                    return i - m + 1
            
            return -1
        
        return searchSubstring(haystack, needle)
